### ISSUE TITLE
System.Reflection.Emit.AssemblyBuilder.Save

### ISSUE THREAD

**Issue by jonorossi** [dotnet/runtime#15704]
`AssemblyBuilder.Save` and `AssemblyBuilderAccess.RunAndSave` isn't available in .NET Core, however coreclr seems to have the code to implement it but I looks conditionally compiled out.

https://github.com/dotnet/coreclr/blob/bc146608854d1db9cdbcc0b08029a87754e12b49/src/mscorlib/src/System/Reflection/Emit/AssemblyBuilder.cs#L1690-L1711

Our use case in Castle DynamicProxy is to write out dynamically created assemblies to disk so we can run peverify over the assembly in unit tests. It also greatly helps writing out the assembly and opening it in ildasm to manually verify IL.

Is there any chance we can get this functionality that .NET Framework has introduced into .NET Core, or an alternative way of doing the same thing. This might not sound like a big deal if you aren't familiar with Castle DynamicProxy's internals, but it really is as it heavily reduces our confidence in both our code and the .NET runtime as well as reducing our ability to track down defects.

**Comment by steveharter** [comment]
Based on discussion with @AArnott and @GrabYourPitchforks this is not a must-have feature for 5.0. However, this feature is completely valid and important but since it is not on the roadmap for 5.0 this is being moved to Future.

**Comment by jmglogow** [comment]
I'm currently porting LibreOffice (LO) to Windows Arm64. LO offers a controlling API / RPC service with various bindings, one of it .NET / C# on Windows. Currently it's based on .NET Framework 4 with a C++/CLI implementation + generated assembly code.

The only option for a native implementation on Arm64 is - AFAIK - to use .NET 5.
The LO API is described in an IDL with an extensible type system, so there are binding specific tools to convert LO's IDL into "code". For .NET this tool is called climaker, which uses this .NET interface to generate the assembly code (https://api.libreoffice.org/docs/tools.html#climaker / https://github.com/LibreOffice/core/tree/master/cli_ure).

The port is in early development and rather experimental, so I currently disabled building the .NET bindings for Arm64, but LO will need a way to generate assembly code again. Or implement some completely new approach with dynamic generated runtime assemblies. All the other approaches I tried, didn't work, but I have never used .NET before and might simply be missing some info to implement a different working solution.

FWIW: more information on the port, including my various approaches on using .NET 5, is in https://bugs.documentfoundation.org/show_bug.cgi?id=137143
P.S. LO has it's own build system and could probably integrate ILPack, either by calling MSBuild or integrating the sources in the build, but I would like to avoid that.

**Comment by weltkante** [comment]
There are different libraries (including third party) that can be used to generate IL code by generators, you don't have to use the reflection-emit API which is limited through various historic constraints in what it can actually emit. Especially if you do offline IL generation as part of the build step I wouldn't chose reflection-emit, this API is for generating executable code into the current process.

The modern framework API for writing offline metadata is in [System.Reflection.Metadata](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.ecma335.metadatabuilder?view=net-5.0) which is what compile-time tools should be using.

**Comment by isral** [comment]
> 
> The modern framework API for writing offline metadata is [MetadataBuilder in System.Reflection.Metadata](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.ecma335.metadatabuilder?view=net-5.0) which is what compile-time tools should be using.

I try to create a compiler (for hobby, not job).
I use reflection emit. I decided to migrate to .NET 5 only later realize I can't save assembly.

There is no documentation on MetadataBuilder. MSDN only show method's parameter list without explanation at all.
I can't find tutorial on MetadataBuilder.

Do you have any suggestion?

**Comment by tomhicklingnc** [comment]
@isral I have used https://github.com/Lokad/ILPack before in .NET Core projects, it's very easy to use and has worked fine for what I needed it for.

**Comment by jinek** [comment]
Guys, just to understand. Isn't it possible to save assembly in dotnetcore currently out of the box? It's just I've been hitting missing parts since 1.0.
I also don't understand the problem. .NET compiler works somehow both on Windows and Linux.

**Comment by AArnott** [comment]
@jinek The problem is that existing .NET APIs under System.Reflection.Emit that can be used to generate dynamic assemblies works on .NET Core except for saving the dynamic assembly to disk. No 3rd party solution addresses this problem as it would require rewriting all your Ref.Emit code to use the new library. Even if this could be done, The Ref.Emit APIs are exclusive in being able to add members to a dynamic assembly after it has already been loaded.

**Comment by isral** [comment]
@jinek 
.net (core) 5 also doesn't have emit methods related to debug information.
https://github.com/Lokad/ILPack mentioned above also can not set entry point.
If you don't need entry point and debug information, give ILPack a try.

**Comment by davidfowl** [comment]
This would be super useful for debugging, for my usage, it's used to dump the IL to disk and open it in ILSpy 😄

**Comment by steveharter** [comment]
Requirements gathered from this issue, with attempt at prioritization:

P1:
- Add `AssemblyBuilder.Save` and the corresponding enum value `AssemblyBuilderAccess.RunAndSave`.
- Support EnC (Edit-and-Continue) meaning the ability to add a new type and have that metadata delta transferred to the already-loaded assembly.

P2:
- Support all platforms (not just Windows).
- Add APIs to emit debugging\symbolic information.
- Add `AssemblyBuilder.SetEntryPoint`.

**Comment by jkotas** [comment]
> EnC 

This is internal implementation detail from AssemblyBuilder.Save point of view. I fully agree that it would be a very good idea to build Reflection.Emit as a layer over EnC , but I would not call it a requirement.

> Add APIs to emit debugging\symbolic information.

We have requests to add symbols in other issues. Symbols are more important than AssemblyBuilder.Save for some of the users.

> Support all platforms (not just Windows).

It would look pretty bad to have Ref.Emit Save on Windows only given our cross platform promise.

I would add more requirement: Ability to use Reflection.Emit against reference assemblies. Otherwise, we will end up with very bad versioning problems when Reflection.Emit is used to build compilers. These versioning problems turn into migration blockers to new .NET Runtime and SDK versions. We had this in .NET Framework and it was not pretty.

**Comment by steveharter** [comment]
Thanks @jkotas  for the feedback. I updated the requirements [above](https://github.com/dotnet/runtime/issues/15704#issuecomment-730528965).

**Comment by steveharter** [comment]
If anyone is truly blocked by the issue for 6.0, please describe the scenario and what requirements from  https://github.com/dotnet/runtime/issues/15704#issuecomment-730528965 are needed, and what is missing.

This is a very large feature and we need to justify the cost : benefit for 6.0.

**Comment by perfolenta** [comment]
My project (compiler) is based on Reflection.Emit and I cannot translate it to Net 5 due to the lack of the Save method. 
I don't think I'm the only one. 
There is no strength to rewrite the entire project, because wrote it for several years.
I hope that at least in 6.0 it will be possible to save a ready-to-run assembly to disk.

**Comment by slozier** [comment]
In the IronPython project, it's used to save compiled Python code to an assembly (to run and/or debug). However it also relies on `LambdaExpression.CompileToMethod` (https://github.com/dotnet/runtime/issues/20270) which was never brought over from .NET Framework so it's unclear if/when it'd be able to use `AssemblyBuilder.Save` again.

**Comment by isral** [comment]
>     * Add `AssemblyBuilder.Save` and the corresponding enum value `AssemblyBuilderAccess.RunAndSave`.

Is it possible to have collect and save?
For language server that compile every user type, is collect needed here?

**Comment by ahartmanns** [comment]
> If anyone is truly blocked by the issue for 6.0, please describe the scenario and what requirements from [#15704 (comment)](https://github.com/dotnet/runtime/issues/15704#issuecomment-730528965) are needed, and what is missing.

We use Reflection.Emit in a [simulation tool](https://www.modestchecker.net/) to compile models into IL that implements a specific interface, which our tool then calls to do the simulation runs. The lack of `AssemblyBuilder.Save` currently blocks the following for us:

* Debugging during development: We cannot save the IL we generate to file in order to run ildasm or peverify equivalents on it. This is a major nuisance that slows down our development significantly, but not a blocker.
* Distributed simulation: To run on multiple machines or a cluster, we compile the code on the master machine, write it to file using AssemblyBuilder.Save, read the bytes of the file back, transfer them to the slave machines, and load the assembly there (via `Assembly.Load(byte[])`). Then every machine can simulate the same model code and transmit results back to the master. Lack of `AssemblyBuilder.Save` completely blocks this scenario; we currently need to revert back to an old version running on .NET Framework/Mono for distributed simulation.

Thus what we'd need from https://github.com/dotnet/runtime/issues/15704#issuecomment-730528965 is support for `AssemblyBuilder.Save` and `AssemblyBuilderAccess.RunAndSave` on all platforms (Windows, macOS, Linux).

We don't _need_ debug info etc., but it could be helpful for the development part (if it gets us nice variable names in ildasm, for example). We tried ILPack (some time ago), but it was buggy and incomplete (at that time); we generate apparently "complex" code with generic types parameterised by value types with interface constraints on interfaces that are in our main assembly (not emitted) etc. that appeared to trigger lots of scenarios not expected by ILPack.

**Comment by jmglogow** [comment]
I was told to port the [LibreOffice's climaker](https://github.com/LibreOffice/core/tree/master/cli_ure/source/climaker) (just ~2.5k LOC) code to [MetadataBuilder in System.Reflection.Metadata](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.ecma335.metadatabuilder?view=net-5.0) in this thread, instead if using this API. I found the documentation a btt lacking with quite a few "Blob" parameters. I'm not really proficient with .NET, so this might be mainly my problem. But it seems that a lot of (equivalent) helper classes, like Emit::FieldBuilder or Emit::ConstructorBuilder, aren't available in the MetadataBuilder API. In the end I came to the conclusion, the result would be a lot like re-implementing ILPack; now having to account for a lot more stuff myself, which was previously done by the AssemblyBuilder interface. Also stuff like creating a type derivated from some existing System type seem rather tedious and the ILPack code also seems to explicitly ignore some cases (I can't remember exactly - this was two months ago).

Maybe a "less flexible, but easier" API on top of MetadataBuilder would help, if people are assumed to use this API now for their previous System.Reflection.Emit.AssemblyBuilder.Save use cases.

I guess at some point someone (me?) has to bite the bullet and do the .NET 5.0 port, not just for Arm64, but currently I decided I don't have the time and it's not clear, if this is really the best way implement a new climaker (also as a DLL run by the dotnet.exe and it uses LibreOffice C++ libraries, so already depends on native code).

**Comment by jonorossi** [comment]
As stated in the issue description:

> Our use case in Castle DynamicProxy is to write out dynamically created assemblies to disk so we can run peverify over the assembly in unit tests. It also greatly helps writing out the assembly and opening it in ildasm to manually verify IL.

With .NET Core/5 getting new runtime features that aren't in .NET Framework we've skated by testing some of them using their underlying representation in IL via attributes and other metadata, but we are not performing the same testing using the .NET Core/5 runtime that we are with .NET Framework's runtime.

We've got just about all the mocking libraries and other projects including Entity Framework depending on Castle DynamicProxy, but as the years pass the .NET Core/5 implementation diverges from .NET Framework so the bar to make changes to DynamicProxy continues to gets higher to remain confident in the generated IL.

@steveharter if you've got any specific questions, I'm happy to answer them.

**Comment by masonwheeler** [comment]
> My project (compiler) is based on Reflection.Emit and I cannot translate it to Net 5 due to the lack of the Save method.
> I don't think I'm the only one.

You're not.  This is a very noticeable omission in the API for exactly that reason.

**Comment by bartdesmet** [comment]
We've been affected by the lack of `Save` as well (and `CompileToMethod` in expression trees) for a high-density large scale event processing system. I'll sketch the basics of it.

The service hosts standing queries which are shipped from clients to services in the form of some query language that gets translated down to .NET expression trees which get compiled and evaluated to instantiate the query. A typical service replica hosts a few 1000s of these, and each service host process runs 100s is service replicas, so we're seeing up to a million queries per process. These queries can run for seconds to several years (literally!).

To achieve reliability, we employ a (differential) checkpointing strategy whereby the queries (as well as runtime state, e.g. a query may compute an average which requires storing a sum and a count) get persisted such that they can be rehydrated upon a failover onto a different node in the cluster. The base implementation of this stores the expression trees in a normalized form using an expression tree serializer framework we built.

From a performance point of view, deserializing and recompiling expression trees is quite costly and is a significant cost for recovery (it involves a lot of object allocation, going through expensive reflection code paths to type check the expression trees behind the factory methods, invoking compilation, and then the whole native side of JIT). This gets mitigated in a few ways, including detecting common subexpressions, generating templates by hoisting out constants, etc.

However, to get to the next level of performance, we built an "always on" system that leverages saving compiled expression trees to disk (as well as auxiliary types built using Reflection.Emit). It's based on a generational assumption akin to garbage collectors. As a standing query gets checkpointed (say every minute), its age is increased. When it reaches a certain age and a certain number of queries have accumulated in an age group, we go through the motions of compiling all of these into an assembly. Upon recovery, we simply reload the assemblies of these older generations (with a goal of it covering > 95% of the queries), and only recover the youngest generations using expression tree deserialization.

For older queries that got deleted, we keep a tombstone table (used to prevent recovering them), and when that table exceeds a certain percentage, we compact the generation (we still save the original expression trees to cold storage, so we can get them back and apply new rounds of common subexpression elimination with newer ones, prior to recompiling them into a MethodBuilder) and coalesce it with younger generations (so upon a future recovery, the smaller compacted assembly gets loaded instead).

In this scheme, recovery is simply loading a few assemblies, discovering all query methods in it (using reflection), excluding the tombstoned items, and invoking the methods that instantiate the query operator graphs. It cuts out all off the expression deserialization, factory invocations, reflection emit, etc. There's an obvious tradeoff in that these assemblies cannot be unloaded so queries that get deleted remain resident (as IL as well as JITted code), but queries tend to live long, create/delete churn is limited, and services rebalance frequently causing the primary replicas to move to other hosts, causing a recovery. Each such recovery will see new assemblies due to the collection and compaction logic we run periodically in the background, so the alive/dead ratio is kept at bay.

The key missing pieces to re-enable this mode are `Save` and `CompileToMethod`. Debug info generation would be a nice to have given that we did implement an expression tree instrumentation mechanism using `DebugInfo` nodes to emit sequence points against a `ToString`/`DebugView` like pretty-printed form of expression trees. We have used this in certain environments to store a `.exp` expression tree "human-readable code file" alongside these persisted compiled expressions, so we can get some symbolic info (file, line, column) when something blows up inside a query expression being evaluated.

**Comment by bartdesmet** [comment]
One more thing. Just a few days ago I did have to fall back to .NET 4.8 to figure out why a piece of Reflection.Emit generated code was causing an `InvalidProgramException`. The inability to `Save` the assembly to disk and run it through `PEVerify.exe` or open it in ILSpy was what caused me to switch back temporarily. (This was for a different project as the one mentioned above.)

**Comment by leppie** [comment]
Author of IronScheme. I use Save to persist libraries to disk (at almost zero effort to me). RE plays a dual role as IronScheme is an incremental compiler. Also it would be silly to have another backend to emit IL, they are never going to be the same, nevermind the huge effort.

**Comment by mjsabby** [comment]
@steveharter do you know why this would be very large? Is it the sum of all parts need to be delivered in one release? Could we not bring back the implementation that existed?

**Comment by jkotas** [comment]
The implementation that existed does not fulfill these requirements from https://github.com/dotnet/runtime/issues/15704#issuecomment-730538328:
- Support all platforms (not just Windows).
- Ability to use Reflection.Emit against reference assemblies to help prevent versioning issues in certain scenarios

**Comment by mjsabby** [comment]
Which all reference assemblies? The ones that ship in box?

Has a design been thought of for the second point? That seems like the most work at least in my head.

To be able to do that we'd need some way of being able to take live-types and walk their dependencies in a different load context where the ref assemblies are? And if we say no that's not possible then users would have change load these types in the metadata load context where the ref assemblies would be, and while it would still be a lot less work by the developers, it wouldn't be as turn key.

Bringing back the old code has the benefit of it works, but doesn't solve the versioning problem which admittedly is a lot worse on .NET Core/.NET X+ although it is something of value if people care. And the Windows-only part (PDBs I imagine?) could be made optional or something.

Yeah no good answer, just thinking out loud to see if there would be any interest in these lesser solutions.

**Comment by masonwheeler** [comment]
> The implementation that existed does not fulfill these requirements from [#15704 (comment)](https://github.com/dotnet/runtime/issues/15704#issuecomment-730538328):
> 
> * Support all platforms (not just Windows).
> * Ability to use Reflection.Emit against reference assemblies to help prevent versioning issues in certain scenarios

Fair enough.  But *it did actually work,* within the parameters of what it was designed to do.  That makes it a good starting point.  Could that code be added in, perhaps on a branch, to give the community something to iterate on?

**Comment by jkotas** [comment]
The code is included in https://github.com/dotnet/coreclr/commit/ef1e2ab328087c61a6878c1e84f4fc5d710aebce. You can cherry pick it from there if you believe that it has future. I believe that trying to iterate on that code is a dead end.

**Comment by steveharter** [comment]
To recap the current status for those wanting to use AssemblyBuilder.Save():

**Currently there are no concrete plans to add this for 6.0; currently evaluating cost:benefit.**

**Workaround include:**
- Continue to use .NET Framework (desktop).
- For static scenarios where compilation can occur ahead-of-time:
  - Roslyn.
    - Supports outputting debugging information (PDB files).
    - Since used internally by the C# compiler, supports [newer language features](https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md).
    - An area of continuing investment including [Source Generators](https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.md).
  -   System.Reflection.DispatchProxy
- https://github.com/Lokad/ILPack
  - A recent [PR ](https://github.com/Lokad/ILPack/pull/129)unblocked ILPack for @AArnott diagnostic scenarios.
  - Diagnostic tools can then be run on this saved IL including:
    - [PEVerify ](https://github.com/dotnet/runtime/tree/master/src/coreclr/src/tools/ILVerify/README.md)
    - https://github.com/icsharpcode/ILSpy
  - Dynamic debugging \ decompile is also possible with third-party tools.

**Comment by AArnott** [comment]
@steveharter Perhaps updated your guidance above to replace PEVerify with [ILVerify](https://github.com/dotnet/runtime/blob/739c53e737e25a47df3d75e39d167463e6da0688/src/coreclr/src/tools/ILVerify/README.md). It turns out PEVerify only works on .NET Framework dll's. For .NET Core dll's, ILVerify works. See https://github.com/Lokad/ILPack/issues/131 for more.

**Comment by AqlaSolutions** [comment]
This is very sad, I thought .NET Core is already mature enough to replace .NET 4.8 but it only looks like...

**Comment by chkn** [comment]
> Ability to use Reflection.Emit against reference assemblies to help prevent versioning issues in certain scenarios

I'm not sure I understand this. Isn't .NET 5/6 supposed to be the one .NET? Why will we need reference assemblies?

I'm working on a project that could really benefit from `AssemblyBuilderAccess.RunAndSave` and `LambdaExpression.CompileToMethod`, so I do hope they're restored. Mono has had cross-platform implementations of these for decades.

**Comment by RomainHautefeuille** [comment]
Is there any efficient alternative to Save + ILSpy for debugging?
My old workflow was to write the target class in C#, write the equivalent class using Emit and compare both using ILSpy, fixing differences one by one.
I can't think of a way to achieve something similar now, or at least somewhat efficient.

**Comment by tannergooding** [comment]
@jkotas, @steveharter  is this currently blocked because:
1. Roslyn never publicly exposed `PeWriter` and `ILBuilder`
2. No one has determined how `AssemblyBuilder.Save` should be built on top of said `PeWriter` and `ILBuilder` APIs

There are some APIs, like `Regex.CompileToAssembly` and others that are having `[Obsolete]` considered as they aren't implemented on .NET Core and never have been.
All of the APIs being considered have this issue blocking them in common.

If there is something actionable here, I'd like to understand what it is and see if we can reach out to the community to unblock these. Otherwise, we are likely going to end up marking such surface as `Obsolete`.

**Comment by masonwheeler** [comment]
72 upvotes and over 5 years later, people are still making it clear that they want this.  All the objections that have been posted so far look more like implementation details that can be worked around with a bit of creativity than actual blockers.  But in 5 years, has anyone actually started working on it?

This is probably the #1 impediment remaining that's stopping people from migrating projects to .NET Core.  Why don't we have it yet?

**Comment by perfolenta** [comment]
> stopping people from migrating projects to .NET Core.
For my projects, this is a very serious obstacle to the migrating to .NetCore or to .Net5... 
It doesn't even matter to me in what format the assembly will be saved. 
The main thing is that it could be launched for execution in some way.

**Comment by MSDN-WhiteKnight** [comment]
>is this currently blocked because: 
>Roslyn never publicly exposed PeWriter and ILBuilder

Why does this need PeWriter from Roslyn? I think it could use ManagedPEBuilder from System.Reflection.Metadata.

>There are some APIs, like Regex.CompileToAssembly and others that are having [Obsolete] considered as they aren't implemented on .NET Core and never have been.

If i understand correctly, the proposed API would be out-of-band package depending on S.R.Metadata. APIs from standard libraries  like Regex could not depend on it. So you could obsolete these regardless of the outcome of this proposal.

**Comment by MSDN-WhiteKnight** [comment]
I think there's another open question, what to do with emitting executables? .NET Framework AssemblyBuilder allowed to generate executables by invoking SetEntryPoint method. This produced single exe file, runnable as is (app.config is optional). But in .NET Core it is more complicated, executable should either have runtimeconfig.json if it runs on shared host, or have embedded host. Should the proposed API deal with these complexities? Or the first version could only support libraries? Probably emitting libraries is more common scenario.

**Comment by perfolenta** [comment]
Out of despair, I will agree to any option. If only libraries are possible, let there be only libraries. :(

**Comment by kg** [comment]
> To recap the current status for those wanting to use AssemblyBuilder.Save():
> 
> **Currently there are no concrete plans to add this for 6.0; currently evaluating cost:benefit.**
> 
> **Workaround include:**
> 
>     * Continue to use .NET Framework (desktop).
> 
>     * For static scenarios where compilation can occur ahead-of-time:
>       
>       * Roslyn.
>         
>         * Supports outputting debugging information (PDB files).
>         * Since used internally by the C# compiler, supports [newer language features](https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md).
>         * An area of continuing investment including [Source Generators](https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.md).
>       * System.Reflection.DispatchProxy
> 
>     * https://github.com/Lokad/ILPack
>       
>       * A recent [PR ](https://github.com/Lokad/ILPack/pull/129)unblocked ILPack for @AArnott diagnostic scenarios.
>       * Diagnostic tools can then be run on this saved IL including:
>         
>         * [PEVerify ](https://github.com/dotnet/runtime/tree/master/src/coreclr/src/tools/ILVerify/README.md)
>         * https://github.com/icsharpcode/ILSpy
>       * Dynamic debugging \ decompile is also possible with third-party tools.

While this provides adequate alternatives for many use cases, this doesn't address LINQ CompileToMethod and it's not obvious how it could be addressed without re-enabling the API. Is the intended alternative to CompileToMethod: do offline Roslyn compiles for everything and ditch LINQ entirely, using a separate code path if you also need to interact with LINQ-consuming APIs like Entity Framework? Is it intended that a user-authored library could reimplement LINQ CompileToMethod much like ILPack, and issues should be filed for any cases where user code can't access the necessary information to compile the expression?

Having to reimplement the whole Compile infrastructure from scratch and have it diverge from what native .Compile does is an unpleasant thought. Should we just go spelunking into private members and pull out the IL generated by Compile and modify it so it can be saved into an assembly?

**Comment by jeremywphelps** [comment]
There are clearly many use cases for generating dynamic assemblies and persisting those to disk. As others have mentioned above, verification and debugging aren't the only ones, though they are very important, Reuse, shadow copying, and performance are others. I'm currently involved in two very large projects that need to do this dynamically from custom metadata, and requirements to do this in .NET Core. 

I think there is a very basic and fundamental need to generate loadable, sharable, and runnable assemblies, using the targeted .NET Core version.

**Comment by IanKemp** [comment]
While this is not a common use-case, it's a very important one. Votes are not the only way that Microsoft should be judging cost:benefit analysis, I can guarantee you that there are many large organizations with thousands of developers between them that will remain stuck on Framework until or unless .NET supports this functionality.

The goal of .NET should be to cover (as far as possible) the existing API surface of Framework **first**, and adding new APIs a very distant second. Get your priorities straight, Microsoft.

**Comment by DalekBaldwin** [comment]
> Out of despair, I will agree to any option. If only libraries are possible, let there be only libraries. :(

In my organization there are already several other scenarios where we can't produce a final artifact directly as an executable for one reason or another, so we create a simple wrapper project with a boilerplate `Main` entry method that just calls the "real" `Main` in the "library". So even this would be tolerable and maintainable until a better option is available.

**Comment by steveharter** [comment]
Update for v7: I'm taking a look at options including an adapter from an [AssemblyBuilder](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.assemblybuilder?view=net-6.0) instance to a [MetadataBuilder](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.ecma335.metadatabuilder?view=net-6.0) instance. I'll report back on this when I have an update.

For now, these sections in this issue are used for tracking:
[requirements](https://github.com/dotnet/runtime/issues/15704#issuecomment-730528965)
[workarounds](https://github.com/dotnet/runtime/issues/15704#issuecomment-732330853)

**Comment by zgramana** [comment]
IKVM also needs this in order to support NET 6+ with the same level of functionality as 4.5+. (Symbol support as well)

**Comment by chkn** [comment]
@masonwheeler After a cursory look at `MetadataBuilder`, I think https://github.com/jbevain/cecil is probably a much better option. It's well established and the de facto standard used by tons of projects. It's also very good at roundtripping assemblies and has a very intuitive/Googlable API.

But, we still need Reflection.Emit.

**Comment by masonwheeler** [comment]
@chkn Can Cecil be used to generate new assemblies from scratch?  Its project page mentions analysis and editing of existing assemblies, but not creation.

**Comment by chkn** [comment]
@masonwheeler yes, just new up a `ModuleDefinition` and add `TypeDefinition`s to it.

**Comment by Luiz-Monad** [comment]
I came here because I wanted to save some assemblies from an AppDomain to debug, then I realized that VisualStudio2019 decompiler can walk over the memory structures and produce source code. Well, at least now you can debug your dynamic emitted code without saving the PDB.

**Comment by danmoseley** [comment]
@steveharter this seems to have been superseded by https://github.com/dotnet/runtime/issues/62956, right? If so let's close this one so we don't have two.

**Comment by steveharter** [comment]
Moving to 8. There is active prototyping work in https://github.com/dotnet/runtimelab/tree/feature/assembly-builder-save.

**Comment by Dangercoder** [comment]
> Moving to 8. There is active prototyping work in https://github.com/dotnet/runtimelab/tree/feature/assembly-builder-save.

Very thankful for the work that is being done .

The removal of AssemblyBuilder.Save() broke clojure-clr's ability to generate standalone DLL's: https://clojure.atlassian.net/browse/CLJCLR-109
https://github.com/clojure/clojure-clr

The clojure-clr users will be very happy ❤️

**Comment by csharper2010** [comment]
> Moving to 8. There is active prototyping work in https://github.com/dotnet/runtimelab/tree/feature/assembly-builder-save.

Hope it will make it to 8 as it would make our path from .NET framework smoother

**Comment by buyaa-n** [comment]
> Is this already included inside the .Net 8.0 Preview 2?

Current status:

- Design done including refactoring Initial implementation for simple assembly
- API review done for Save(): https://github.com/dotnet/runtime/issues/83988
- Initial implementation for simple assembly is in [PR](https://github.com/dotnet/runtime/pull/83554)

We plan to have MVP (Minimum Viable Product) implementation done for preview 4, read [details about what it would have](https://github.com/dotnet/runtime/issues/62956). After MVP implementation we would mostly rely on community contributions for completing this work

**Comment by buyaa-n** [comment]
Update: We have added partial support for AssemblyBuilder.Save in .NET 8:

We have completed our goal to add MVP (Minimum Viable Product) support in .NET 8 that allows us to save *.tlb files produced by [TlbImp tool](https://learn.microsoft.com/en-us/dotnet/framework/tools/tlbimp-exe-type-library-importer). Now we can port the tool into .NET Core using the new AssemblyBuilder.Save implementation.

The implementation support following:
- Create persist able AssemblyBuilder with or without custom attributes
- Define ModuleBuilder (same as existing AssemblyBuilder only one module supported)
- Define classes, interfaces, structs
- Define fields, asbstract methods, method parameters for them. 
- Add custome attributes for each member
- Save the assebmly

The implementation doesn't support other members (a method with body (IL not supported), Properties, Constructors, Events). Because we have only partial implementaion we did not exposed [the approved new APIs](https://github.com/dotnet/runtime/issues/83988#issuecomment-1487355324 publicly). Those APIs are internal, if you want to give it a try you could use reflection, example:

```cs
using System.Reflection.Emit;
using System.Reflection;

namespace TestApp
{
    internal class Program
    {
        static void Main(string[] args)
        {
            AssemblyName aName = new AssemblyName("MyAssembly");
            AssemblyBuilder ab= PopulateAssemblyBuilderAndSaveMethod(aName, null, out MethodInfo saveMethod);
            ModuleBuilder moduleBuilder = ab.DefineDynamicModule("ModuleName"); // Only one module supported
            moduleBuilder.DefineType("TestInterface", TypeAttributes.Interface | TypeAttributes.Abstract);
            // Add other classes, interfaces, and structs. Define method, field, custom attributes for them

            // Note that non abstract method not supported yet as writing IL is not supporeted yet
            // defining Constructors, Properties and Events are also not supported yet

            // Then save the assembly into a file (or Stream)
            saveMethod.Invoke(ab, new object[] { "MyAssembly.dll" });
        }

        private static AssemblyBuilder PopulateAssemblyBuilderAndSaveMethod(AssemblyName assemblyName,
                List<CustomAttributeBuilder>? assemblyAttributes, out MethodInfo saveMethod)
        {
            Type abType= Type.GetType("System.Reflection.Emit.AssemblyBuilderImpl, System.Reflection.Emit", throwOnError: true)!;

            saveMethod = abType.GetMethod("Save", BindingFlags.NonPublic | BindingFlags.Instance,
                                         new Type[] { typeof(string) /* or use typeof(Stream) */ });

            MethodInfo defineDynamicAssembly = abType.GetMethod("DefinePersistedAssembly", BindingFlags.NonPublic | BindingFlags.Static,
                new Type[] { typeof(AssemblyName), typeof(Assembly), typeof(List<CustomAttributeBuilder>) });

            return (AssemblyBuilder)defineDynamicAssembly.Invoke(null, new object[] { assemblyName, typeof(object).Assembly, assemblyAttributes });
        }
    }
}
```

**Comment by ForNeVeR** [comment]
While this is, of course, some good news, the implemented surface is still very far from what we would ideally like to see.

In most use cases I've seen so far, people wanted to save assemblies with actual _executable code_ / method bodies, and as far as I understand, this is yet not implemented.

**Comment by MichaelKetting** [comment]
@ForNeVeR I got the definitive answer here in https://github.com/dotnet/runtime/issues/62956#issuecomment-1645654495

Summary: No, right now we should not expect support executable code / method bodies in .NET 8

**Comment by buyaa-n** [comment]
UPDATE: we are making a good progress:
- [x] Added ILGenerator implementation

- [x] Added support for remaining members
  - [x] ConstructorBuilderImpl
  - [x] PropertyBuilderImpl
  - [x] EventBuilderImpl
- [x] Added implementations for other unimplemented APIs in AssemblyBuilderImpl, TypeBuilderImp, MethodBuilderImpl
  - [ ] Still need to add a few APIs like `CreateGlobalFuntions`, `DefineGlobalMethod` etc. The work is in progress

Planning to finish the main functionality (excluding entry point and PDB) and [make related APIs public](https://github.com/dotnet/runtime/issues/83988) in preview 1. Here is the tracking issue for details: https://github.com/dotnet/runtime/issues/92975.

Now it would be great to test the implementation with a real-life scenario. I would really appreciate if you could:
- Prepare a simple app with your use case and share with me (in this thread or email me)
- Dogfood the [daily build](https://github.com/dotnet/installer#installers-and-binaries) and try out the new AssemblyBuilder implementation yourself and give feedback. (The APIs are still internal so need to use reflection for now as [mentioned in the sample](https://github.com/dotnet/runtime/issues/15704#issuecomment-1644571703))

Thank you in advance!

**Comment by kg** [comment]
> Now it would be great to test the implementation with a real-life scenario. I would really appreciate if you could:
> 
>     * Prepare a simple app with your use case and share with me (in this thread or email me)
> 
>     * Provide a sample assembly similar what your app generates, we can use for testing.

I have a large application currently targeting net4x that uses AssemblyBuilder. Would having generated assemblies from it be useful at all for this? I could also provide the code for the implementation, but I don't have the spare time to do a test port to NET8/NET9 yet. It's a fairly complex script compiler with a small automated test suite, and then there are around a hundred scripts or so that are loaded by a host .NET application. It leans heavily on LINQ Expression.CompileToMethod to generate most of the methods (and does classic SRE for things that LINQ can't generate, like instance methods or pointers), so I don't know if that is relevant here and making that scenario work is considered in scope.

**Comment by buyaa-n** [comment]
> I have a large application currently targeting net4x that uses AssemblyBuilder. Would having generated assemblies from it be useful at all for this?

Yes, the generated assembly with your app would be useful, thanks!

> It leans heavily on LINQ Expression.CompileToMethod to generate most of the methods (and does classic SRE for things that LINQ can't generate, like instance methods or pointers), so I don't know if that is relevant here and making that scenario work is considered in scope.

LINQ `Expression.CompileToMethod` is not relevant here, though the assembly created with combination of LINQ Expression.CompileToMethod and SRE could be useful for testing

**Comment by acaly** [comment]
Will `XmlSerializer.GenerateSerializer` be added back, or we will only have the Microsoft.XmlSerializer.Generator package?

**Comment by jkotas** [comment]
> Will XmlSerializer.GenerateSerializer be added back, or we will only have the Microsoft.XmlSerializer.Generator package?

I do not expect that XmlSerializer.GenerateSerializer be added back. See #1413 and related issues.

**Comment by buyaa-n** [comment]
>> I have a large application currently targeting net4x that uses AssemblyBuilder. Would having generated assemblies from it be useful at all for this?
>>
>Yes, the generated assembly with your app would be useful, thanks! 
>I have a testing tool that would try to regenerate the assembly with .NET Core implementation

Thank you @kg for sending assemblies, I have regenerated the assembly with .NET Core implementation and found [many issues](https://github.com/dotnet/runtime/pull/97350) that now [fixed](https://github.com/dotnet/runtime/pull/98393) and will be ported with .NET 9 preview 2.

**Comment by buyaa-n** [comment]
> Planning to finish the main functionality (excluding entry point and PDB) and [make related APIs public](https://github.com/dotnet/runtime/issues/83988) in preview 1.

The feature is now available in https://dotnet.microsoft.com/en-us/download/dotnet/9.0, please give it a try and give us a feedback. We also adding bunch of fixes in preview 2 mostly related to complex generic types/methods

**Comment by mjsabby** [comment]
Will Portable PDB support be there as well in this preview?

**Comment by buyaa-n** [comment]
Entry point and PDB not included in preview 1. Entry point will be added by preview 3. PDB support timeline not clear yet.

**Comment by jgh07** [comment]
There are a few methods still missing unrelated to PDB generation:

- AssemblyBuilder.DefineUnmanagedResource
- AssemblyBuilder.AddResourceFile

There might be alternatives that I'm unaware of. If not, will these methods also be supported in the future?

**Comment by masonwheeler** [comment]
> PDB support timeline not clear yet.

I really hope it makes it in by release.  At its core, this is a tool for compiler developers, and a compiler that can't produce debug builds is severely disadvantaged.

**Comment by Martin1994** [comment]
> this is a tool for compiler developers, and a compiler that can't produce debug builds is severely disadvantaged.

Portable PDB is not blocked. `System.Reflection.Metadata.Ecma335` is for this use case (not documented too well though IMHO). See https://github.com/Martin1994/ikvm/blob/18b1fce8265a3f128f4742b506f8e10e0cae0f51/src/IKVM.Reflection/Impl/PortablePdbWriter.cs

`System.Reflection.Emit` is more for dynamic emitting use cases, such as IronPython I believe.

**Comment by masonwheeler** [comment]
@Martin1994 That link looks kind of useful, but it looks like it only works if you already have an existing `System.Reflection.Metadata` emitter.

> `System.Reflection.Emit` is more for dynamic emitting use cases, such as IronPython I believe.

IIRC IronPython style dynamic emitting was covered by the DLR, which was released with .NET Framework 4.0.  `System.Reflection.Emit`, and the ability to for compiler writers to use it to save new assemblies, has been around since 1.0.

**Comment by buyaa-n** [comment]
FYI: We took completely different approach for adding `EntryPoint` support and setting other options, we are providing all metadata information produced with Reflection.Emit APIs (MetadataBuilder and ILStreams) so that user could embed them into the corresponding section of `PEBuidler`. This will let the customers handle their assembly building process by themselves using the [PEHeaderBuilder](https://learn.microsoft.com/dotnet/api/system.reflection.portableexecutable.peheaderbuilder.-ctor) and [ManagedPEBuilder](https://learn.microsoft.com/dotnet/api/system.reflection.portableexecutable.managedpebuilder.-ctor) options. Because the [MetadataBuilder](https://learn.microsoft.com/dotnet/api/system.reflection.metadata.ecma335.metadatabuilder) and [BlobBuilder](https://learn.microsoft.com/dotnet/api/system.reflection.metadata.blobbuilder) types are not accessible from within CoreLib we decided to make the `PersistedAssemblyBuilder` type public and reintroduce the related new APIs there, remove the new APIs from the base `AssemblyBuilder` type.
```diff
namespace System.Reflection.Emit;
public abstract partial class AssemblyBuilder
{
    // Previously approved new APIs
-   public static AssemblyBuilder DefinePersistedAssembly(AssemblyName name, Assembly coreAssembly, IEnumerable<CustomAttributeBuilder>? assemblyAttributes = null);

-   public void Save(Stream stream);
-   public void Save(string assemblyFileName);
-   protected abstract void SaveCore(Stream stream);
}

+public sealed class PersistedAssemblyBuilder : AssemblyBuilder
{
+   public PersistedAssemblyBuilder(AssemblyName name, Assembly coreAssembly, IEnumerable<CustomAttributeBuilder>? assemblyAttributes = null);

+   public void Save(Stream stream);
+   public void Save(string assemblyFileName);
+   public MetadataBuilder GenerateMetadata(out BlobBuilder ilStream, out BlobBuilder mappedFieldData);
}
```
See more details and usage example from the [API proposal](https://github.com/dotnet/runtime/issues/97015). This change will be available in preview 3.

**Comment by buyaa-n** [comment]
Further the PDB support will be handled similar way as one would set `EntryPoint` in `PEBuidler`, here is the initial [API proposal](https://github.com/dotnet/runtime/issues/99935) and usage example, please give your feedbacks.

With that I am closing this issue as alternate for the `AssemblyBuilder.Save` is implementation is done
